File: codegen/codegen-erlang/src/test/resources/ir/http_dispatch_module.expected.erl
%% Generated HTTP dispatcher for smithy.beam.demo.http#HttpService.
%% Uses httpc from OTP. Replace via adapter for testing.
-module(runtime_http).
-include("runtime_types.hrl").
-export([dispatch/2, dispatch/3]).

dispatch(Config, Request) ->
    HttpClient = maps:get(http_client, Config, httpc),
    dispatch(HttpClient, Config, Request).

dispatch(HttpClient, Config, Request) -> dispatch_signed(HttpClient, Config, Request).

dispatch_signed(HttpClient, Config, #http_request{
    method = Method, path = Path, query = Query, headers = Headers, body = Body, host = Host
}) ->
    BaseUrl =
        case maps:get(base_url, Config, undefined) of
            undefined ->
                case maps:get(endpoint_prefix, Config, undefined) of
                    undefined -> <<>>;
                    _ -> runtime_helpers:resolve_base_url(Config)
                end;
            GivenUrl ->
                GivenUrl
        end,
    QueryStr =
        case maps:to_list(Query) of
            [] ->
                <<>>;
            Pairs ->
                Encoded = uri_string:compose_query([{K, V} || {K, V} <- Pairs]),
                <<"?", Encoded/binary>>
        end,
    {Scheme, DefaultAuthority} = split_base_url(BaseUrl),
    Authority =
        case Host of
            undefined -> DefaultAuthority;
            _ -> Host
        end,
    ReqUrl = <<Scheme/binary, Authority/binary, Path/binary, QueryStr/binary>>,
    HttpcHeaders = [{binary_to_list(K), binary_to_list(V)} || {K, V} <- Headers],
    Req =
        case Body of
            <<>> -> {binary_to_list(ReqUrl), HttpcHeaders};
            _ -> {binary_to_list(ReqUrl), HttpcHeaders, mime(Headers), Body}
        end,
    case
        HttpClient:request(binary_to_atom(string:lowercase(Method), utf8), Req, [], [
            {body_format, binary}
        ])
    of
        {ok, {{_, Status, _}, RespHeaders, RespBody}} ->
            BinHeaders = [{list_to_binary(K), list_to_binary(V)} || {K, V} <- RespHeaders],
            {ok, #http_response{
                status = Status,
                headers = BinHeaders,
                body = RespBody
            }};
        {error, Reason} ->
            {error, Reason}
    end.

split_base_url(<<>>) ->
    {<<>>, <<>>};
split_base_url(BaseUrl) ->
    case uri_string:parse(binary_to_list(BaseUrl)) of
        #{scheme := Scheme, host := Host} = Parts ->
            PortSuffix =
                case maps:get(port, Parts, undefined) of
                    undefined -> <<>>;
                    Port -> <<":", (integer_to_binary(Port))/binary>>
                end,
            {<<(list_to_binary(Scheme))/binary, "://">>, <<
                (list_to_binary(Host))/binary, PortSuffix/binary
            >>};
        _ ->
            {<<>>, BaseUrl}
    end.

mime(Headers) ->
    case proplists:get_value(<<"Content-Type">>, Headers) of
        undefined -> "application/octet-stream";
        CT -> binary_to_list(CT)
    end.

File: codegen/codegen-elixir/src/test/resources/ir/http_dispatch_module.expected.ex
defmodule RuntimeHttp do
  @moduledoc "Generated HTTP dispatcher for Smithy service clients. Uses Req."
  alias RuntimeTypes, as: RuntimeTypes
  alias RuntimeHelpers, as: RuntimeHelpers

  @spec dispatch(map(), RuntimeTypes.HttpRequest.t()) ::
          {:ok, RuntimeTypes.HttpResponse.t()} | {:error, term()}
  def dispatch(
    config,
    req
  ) do
    http_client = Map.get(config, :http_client, __MODULE__.ReqClient),
    dispatch(http_client, config, req)
  end

  @spec dispatch(module(), map(), RuntimeTypes.HttpRequest.t()) ::
          {:ok, RuntimeTypes.HttpResponse.t()} | {:error, term()}
  def dispatch(
    http_client,
    config,
    req = %RuntimeTypes.HttpRequest{}
  ) do
    dispatch_signed(http_client, config, req)
  end

  @spec dispatch_signed(module(), map(), RuntimeTypes.HttpRequest.t()) ::
          {:ok, RuntimeTypes.HttpResponse.t()} | {:error, term()}
  defp dispatch_signed(
    http_client,
    config,
    req = %RuntimeTypes.HttpRequest{}
  ) do
    base_url =
      case Map.get(config, :base_url) do
        nil ->
          case Map.get(config, :endpoint_prefix) do
            nil -> ""
            _ -> RuntimeHelpers.resolve_base_url(config)
          end
    
        url ->
          url
      end
    
    {scheme, default_authority} = split_base_url(base_url)
    
    authority =
      case req.host do
        nil -> default_authority
        host -> host
      end
    
    url = scheme <> authority <> req.path
    req_opts = [
      method: String.downcase(req.method) |> String.to_atom(),
      url: url,
      params: req.query,
      headers: req.headers,
      body: req.body,
      decode_body: false
    ]
    case http_client.request(req_opts) do
      {:ok, %{status: status, headers: headers, body: body}} ->
        {:ok,
         %RuntimeTypes.HttpResponse{
           status: status,
           headers: Enum.map(headers, fn {k, v} -> {k, v} end),
           body: body
         }}
    
      {:error, reason} ->
        {:error, reason}
    end
  end

  defp split_base_url(""), do: {"", ""}
  defp split_base_url(base_url) do
    case URI.parse(base_url) do
      %URI{scheme: scheme, host: host} = uri when is_binary(host) ->
        port_suffix =
          case {uri.scheme, uri.port} do
            {"https", 443} -> ""
            {"http", 80} -> ""
            {_, nil} -> ""
            {_, port} -> ":#{port}"
          end
    
        {scheme <> "://", host <> port_suffix}
    
      _ ->
        {"", base_url}
    end
  end

  defmodule ReqClient do
    @moduledoc "false"
    @spec request(keyword()) :: {:ok, map()} | {:error, term()}
    def request(req_opts) do
      case Req.request(req_opts) do
        {:ok, %Req.Response{status: status, headers: headers, body: body}} -> {:ok, %{status: status, headers: headers, body: body}}

        {:error, reason} -> {:error, reason}
      end
    end
  end
end